ebos/hdf5serializer.hh

```diff
@@ -0,0 +1,103 @@
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+
+  Consult the COPYING file in the top-level source directory of this
+  module for the precise wording of the license and the list of
+  copyright holders.
+*/
+#ifndef ECL_HDF5_SERIALIZER_HH
+#define ECL_HDF5_SERIALIZER_HH
+
+#include <opm/common/utility/Serializer.hpp>
+#include <opm/common/utility/MemPacker.hpp>
+
+#include <opm/simulators/utils/HDF5File.hpp>
+#include <opm/simulators/utils/moduleVersion.hpp>
+
+namespace Opm {
+
+//! \brief Class for (de-)serializing using HDF5.
+class HDF5Serializer : public Serializer<Serialization::MemPacker> {
+public:
+    HDF5Serializer(const std::string& fileName, HDF5File::OpenMode mode)
+        : Serializer<Serialization::MemPacker>(m_packer)
+        , m_packer{}
+        , m_h5file(fileName, mode)
+    {}
+
+    //! \brief Serialize and write data to restart file.
+    //! \tparam T Type of class to write
+    //! \param data Class to write restart data for
+    template<class T>
+    void write(T& data,
+               const std::string& group,
+               const std::string& dset)
+    {
+        try {
+            this->pack(data);
+        } catch (...) {
+            m_packSize = std::numeric_limits<size_t>::max();
+            throw;
+        }
+
+        m_h5file.write(group, dset, m_buffer);
+    }
+
+    //! \brief Writes a header to the file.
+    //! \param simulator_name Name of simulator used
+    //! \param module_version Version of simulator used
+    //! \param time_stamp Built time-stap for simulator used
+    //! \param case_name Name of case file is associated with
+    void writeHeader(const std::string& simulator_name,
+                     const std::string& module_version,
+                     const std::string& time_stamp,
+                     const std::string& case_name)
+    {
+        try {
+            this->pack(simulator_name, module_version, time_stamp, case_name);
+        } catch (...) {
+            m_packSize = std::numeric_limits<size_t>::max();
+            throw;
+        }
+        m_h5file.write("/", "simulator_info", m_buffer);
+    }
+
+    //! \brief Read data and deserialize from restart file.
+    //! \tparam T Type of class to read
+    //! \param data Class to read restart data for
+    template<class T>
+    void read(T& data,
+              const std::string& group,
+              const std::string& dset)
+    {
+        m_h5file.read(group, dset, m_buffer);
+        this->unpack(data);
+    }
+
+    //! \brief Returns the last group (report step) stored in file.
+    int lastGroup() const
+    {
+        return m_h5file.lastGroup();
+    }
+
+private:
+    const Serialization::MemPacker m_packer; //!< Packer instance
```

Comment:
As the Serializer base class already hold a reference to a packer using the same variable name, there is a risk of confusion. I suggest renaming this m_packer_private or something similar.

Reply:
done